### PR TITLE
feat(common): support injecting custom headers

### PR DESCRIPTION
**Pull request by scotthart**
Adds support for adding custom headers for both gRPC and REST.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13829)

### CODE REVIEW

File: google/cloud/common_options.h [hunk 1]
```diff
@@ -227,12 +228,22 @@ struct EnableServerRetriesOption {
   using Type = bool;
 };
 
+/**
+ * An option to inject custom headers into the request.
+ *
+ * For REST endpoints, these headers are added to the HTTP headers. For gRPC
+ * endpoints, these headers are added to the `grpc::ClientContext` metadata.
```
Comment (dbolduc) [inline google/cloud/common_options.h:235]: add: `@ingroup options` which I am noticing I forgot to do for `EnableServerRetriesOption`.

File: generator/internal/metadata_decorator_generator.cc [hunk 2]
```diff
@@ -406,6 +406,11 @@ void $metadata_class_name$::SetMetadata(grpc::ClientContext& context,
   }
   auto const& authority = options.get<AuthorityOption>();
   if (!authority.empty()) context.set_authority(authority);
+  if (options.has<google::cloud::CustomHeadersOption>()) {
```
Comment (devbww) [inline generator/internal/metadata_decorator_generator.cc:409]: Optional nit:  It looks like we're a little inconsistent with our qualifying.  `Options` a few lines up, for example, is mentioned without `google::cloud::`.

File: generator/internal/metadata_decorator_generator.cc [hunk 3]
```diff
@@ -406,6 +406,11 @@ void $metadata_class_name$::SetMetadata(grpc::ClientContext& context,
...
+    for (auto const& h : options.get<google::cloud::CustomHeadersOption>()) {
```
Comment (dbolduc) [inline generator/internal/metadata_decorator_generator.cc:410]: Drop the `options.has<>` conditional?

Also, s/google::cloud::CustomHeadersOption/CustomHeadersOption/

File: generator/internal/metadata_decorator_rest_generator.cc [hunk 4]
```diff
@@ -400,6 +400,11 @@ void $metadata_rest_class_name$::SetMetadata(
     rest_context.AddHeader("x-server-timeout",
         ms_rep.insert(ms_rep.size() - 3, "."));
   }
+  if (options.has<google::cloud::CustomHeadersOption>()) {
+    for (auto const& h : options.get<google::cloud::CustomHeadersOption>()) {
```
Comment (dbolduc) [inline generator/internal/metadata_decorator_rest_generator.cc:404]: ditto

**Review by dbolduc — APPROVED** [review]

File: generator/internal/metadata_decorator_generator.cc (hunk 2 above)
Comment (devbww) [inline generator/internal/metadata_decorator_generator.cc:409]: Optional nit: You could do without the `has<T>()` test (and hence the double lookup) in this case, at the cost of a once-only allocation of an empty map.

**Review by devbww — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13829)* status: 0 of 427 files reviewed, 5 unresolved discussions (waiting on @dbolduc and @devbww)

*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13829#gh-1536155391:-Ntbzuri6VTJHlstKy0f:bttr68t) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/13829#discussion_r1536155391) to the right spot — sorry!)*

add: `@ingroup options` which I am noticing I forgot to do for `EnableServerRetriesOption`.
</blockquote></details>

Added for all options in this file missing the tag.

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13829#gh-1536158948:-NtbzzAqDGFAGNWV_9dx:bm8aign) (no related file):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/13829#discussion_r1536158948) to the right spot — sorry!)*

Optional nit:  It looks like we're a little inconsistent with our qualifying.  `Options` a few lines up, for example, is mentioned without `google::cloud::`.
</blockquote></details>

Made qualification consistent across all options in the function.

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13829#gh-1536159361:-Ntc-8HA6BjaIQ7TL3_V:b-896fix) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/13829#discussion_r1536159361) to the right spot — sorry!)*

Drop the `options.has<>` conditional?

Also, s/google::cloud::CustomHeadersOption/CustomHeadersOption/
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13829#gh-1536159536:-Ntc-8fLDplMj7mps-r5:b-896fix) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/13829#discussion_r1536159536) to the right spot — sorry!)*

ditto
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13829#gh-1536162433:-Ntc-9fhEq_VjfXRguMM:b-896fix) (no related file):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/13829#discussion_r1536162433) to the right spot — sorry!)*

Optional nit: You could do without the `has<T>()` test (and hence the double lookup) in this case, at the cost of a once-only allocation of an empty map.
</blockquote></details>

Done.

___

**Pull request merged**